File: 3346-maximum-frequency-of-an-element-after-performing-operations-i/3346-maximum-frequency-of-an-element-after-performing-operations-i.cpp
int __FAST_IO__ = []() { std::ios::sync_with_stdio(0); std::cin.tie(0); std::cout.tie(0); return 0; }();

class Solution {
public:
    int maxFrequency(vector<int>& nums, int k, int numOperations) {
        sort(nums.begin(), nums.end());
        int n = nums.size(), ret = 0;
        for (int i = 0, j = 0; i < n; ++i) {
            while (nums[i] - nums[j] > 2 * k) ++j;
            ret = max(ret, min(i - j + 1, numOperations));
        }

        for (int i = 0, l = 0, r = 0; i < n; ++i) {
            int last_i = i;
            while (i + 1 < n && nums[i] == nums[i + 1]) i++;
            while (r < n && nums[r] <= nums[i] + k) ++r;
            while (nums[i] > nums[l] + k) ++l;

            int same = i - last_i + 1;
            int len = r - l;
            ret = max(ret, same + min(len - same, numOperations));
        }
        return ret;
    }
};

File: 3346-maximum-frequency-of-an-element-after-performing-operations-i/3346-maximum-frequency-of-an-element-after-performing-operations-i.py
class Solution:
    def maxFrequency(self, nums: List[int], k: int, numOperations: int) -> int:
        d = set()
        for x in nums:
            d.add(x - k)
            d.add(x)
            d.add(x + 1)
            d.add(x + k + 1)
        d = sorted(list(d))
        n = len(d)
        s = [0] * n
        t = [0] * n
        for x in nums:
            i = bisect_left(d, x - k)
            j = bisect_left(d, x)
            l = bisect_left(d, x + 1)
            m = bisect_left(d, x + k + 1)
            s[i] += 1
            s[j] -= 1

            t[j] += 1

            s[l] += 1
            s[m] -= 1
        ans = 0
        for i in range(n):
            if i:
                s[i] += s[i - 1]
            ans = max(ans, t[i] + min(s[i], numOperations))
        return ans